staging_vespalib/src/vespa/vespalib/util/assert.cpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "assert.h"
+#include <vespa/defaults.h>
+#include <vespa/vespalib/util/backtrace.h>
+#include <fstream>
+#include <map>
+#include <mutex>
+#include <chrono>
+#include <iomanip>
+
+#include <vespa/log/log.h>
+LOG_SETUP(".vespa.assert");
+
+namespace vespalib::assert {
+
+namespace {
+
+std::mutex _G_lock;
+std::map<std::string, size_t> _G_assertMap;
+
+}
+
+size_t getNumAsserts(const char *key)
+{
+    std::lock_guard guard(_G_lock);
+    return _G_assertMap[key];
+}
+
+void assertOnceOrLog(const char *expr, const char *key, size_t freq)
+{
+    std::string relativePath("tmp/");
+    relativePath += key;
+    relativePath += ".assert.";
+    relativePath += vespa::Defaults::vespaUser();
+    std::string rememberAssert = vespa::Defaults::underVespaHome(relativePath.c_str());
+    std::ifstream prevAssertFile(rememberAssert.c_str());
+    if (prevAssertFile) {
+        size_t count(0);
+        {
+            std::lock_guard guard(_G_lock);
+            count = _G_assertMap[key]++;
+        }
+        if ((count % freq) == 0) {
+            LOG(error, "assert(%s) named '%s' has failed %zu times. Stacktrace = %s",
+                       expr, key, count+1, vespalib::getStackTrace(0).c_str());
+        }
+    } else {
+        {
+            LOG(error, "assert(%s) named '%s' failed first time. Stacktrace = %s",
+                       expr, key, vespalib::getStackTrace(0).c_str());
+            std::ofstream assertStream(rememberAssert.c_str());
```

Comment:
A file is opened in a directory with very relaxed permissions. Extra checks are needed to avoid security
issues.  It might be better to introduce another directory with stricter permissions.

Reply:
I am intentionally using VESPA-HOME/tmp as that is often wiped on vespa upgrade. And there is nothing  written to this file that is secret. The file is never read. The existence of a file will do nothing, but prevent a core dump. I see no reason to make this more complicated.

Reply:
A symlink that is not present when constructing prevAssertFile but is present when constructing assertStream
can cause the process to attempt to write to any file that symlink points to.

Reply:
That is highly unlikely.
However I found that we have a tmp directory under var/db/vespa/tmp, so I put it there.
I also added vespa version to the file name so that we will get one core per version.